src/validation_functions.jl

```diff
@@ -0,0 +1,144 @@
+            # Recurse. Components are validated separately and do not need to be validated twice.
+            if !validate_fields(sys, getfield(ps_struct, name))
+                is_valid = false
+            end
+        else
+            field_descriptor = get_field_descriptor(struct_descriptor, string(name))
+            if !haskey(field_descriptor, "valid_range")
+                continue
+            end
+            valid_range = field_descriptor["valid_range"]
+            limits = get_limits(valid_range, ps_struct)
+            valid_info = ValidationInfo(field_descriptor, struct_descriptor["struct_name"],
+                                                              ps_struct, fieldtype, limits)
+            if !validate_range(valid_range, valid_info, field_value)
+                is_valid = false
+            end
+        end
+    end
+    return is_valid
+end
+
+function get_limits(valid_range::String, ps_struct::PowerSystemType)
+    #Gets min and max values from activepowerlimits for activepower, etc.
+    limits = getfield(ps_struct, Symbol(valid_range))
+    return limits
+end
+
+function get_limits(valid_range::Dict, unused::PowerSystemType)
+    #Gets min and max value defined for a field, e.g. "valid_range": {"min":-1.571, "max":1.571}.
+    limits = (min = valid_range["min"], max = valid_range["max"])
+    return limits
+end
+
+function validate_range(::String, valid_info::ValidationInfo, field_value)
+    #Validates activepower against activepowerlimits, etc.
+    is_valid = true
+    if !isnothing(valid_info.limits)
+        is_valid = check_limits_impl(valid_info, field_value)
+    end
+    return is_valid
+end
+
+function validate_range(::Dict, valid_info::ValidationInfo, field_value)
+    return check_limits(valid_info.field_type, valid_info, field_value)
+end
+
+function check_limits(::Type{T}, valid_info::ValidationInfo, field_value) where T <: Union{Nothing, Float64}
+    #Validates numbers.
+    return check_limits_impl(valid_info, field_value)
+end
+
+function check_limits(::Type{T}, valid_info::ValidationInfo, field_value) where T <: Union{Nothing, NamedTuple}
+    #Validates up/down, min/max, from/to named tuples.
+    @assert length(field_value) == 2
+    result1 = check_limits_impl(valid_info, field_value[1])
+    result2 = check_limits_impl(valid_info, field_value[2])
+    return result1 && result2
+end
+
+function check_limits_impl(valid_info::ValidationInfo, field_value)
+    is_valid = true
+    action_function = get_validation_action(valid_info.field_descriptor)
+    if (!isnothing(valid_info.limits.min) && field_value < valid_info.limits.min) ||
+        (!isnothing(valid_info.limits.max) && field_value > valid_info.limits.max)
+        is_valid = action_function(valid_info, field_value)
+    end
+    return is_valid
+end
+
+function get_validation_action(field_descriptor::Dict)
+    action = get(field_descriptor, "validation_action", "error")
+    if action == "warn"
```

Comment:
Shouldn't these actions be on an enum? Or do we want to have string comparisons?

Reply:
Since Julia doesn't support enums with string values we have to implement string comparisons at some level.  I think this is the simplest solution for this case.  The user has to enter "warn" or "error".